

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Description &mdash; mkCOInr  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9a2dae69"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Input/Output" href="io.html" />
    <link rel="prev" title="Tutorial" href="tutorial.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            mkCOInr
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview of mkCOInr</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="run.html">Run mkCOInr</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Description</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#add-taxids-pl">add_taxids.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#aim">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#input-files">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parameters-options">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#algorithm">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#dereplicate-pl">dereplicate.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#download-bold-pl">download_bold.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id6">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id7">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id8">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id9">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id10">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#download-taxonomy-pl">download_taxonomy.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id11">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#input">Input</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id13">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id14">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#format-bold-pl">format_bold.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id15">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id16">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id17">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id18">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id19">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#format-bold-package-pl">format_bold_package.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id20">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id21">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id22">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id24">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#format-custom-pl">format_custom.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id25">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id26">Input</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id27">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id28">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id29">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#format-db-pl">format_db.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id30">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id31">Input</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id32">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id33">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id34">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#format-rdp-pl">format_rdp.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id35">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id36">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id37">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id38">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id39">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#format-ncbi-pl">format_ncbi.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id40">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id41">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id42">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id43">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id44">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#get-subtaxa-pl">get_subtaxa.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id45">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id46">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id47">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id48">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id49">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#pool-and-dereplicate-pl">pool_and_dereplicate.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id50">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id51">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id52">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id53">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id54">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#reduce-metadata-pl">reduce_metadata.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id55">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id56">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id57">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id58">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#select-region-pl">select_region.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id59">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id60">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id61">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id62">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id63">Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#select-taxa-pl">select_taxa.pl</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id64">Aim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id65">Input files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id66">Parameters/options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id67">Algorithm</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id68">Output</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="io.html">Input/Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="reflist.html">List of References</a></li>
<li class="toctree-l1"><a class="reference internal" href="cite.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">mkCOInr</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Description</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/content/reference.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="description">
<h1>Description<a class="headerlink" href="#description" title="Link to this heading"></a></h1>
<section id="add-taxids-pl">
<span id="add-taxids-reference"></span><h2>add_taxids.pl<a class="headerlink" href="#add-taxids-pl" title="Link to this heading"></a></h2>
<section id="aim">
<h3>Aim<a class="headerlink" href="#aim" title="Link to this heading"></a></h3>
<p>For each lineage</p>
<blockquote>
<div><ul class="simple">
<li><p>Find the smallest taxon that matches an already existing <a class="reference internal" href="glossary.html#taxid-glossary"><span class="std std-ref">taxID</span></a>  in <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a></p></li>
<li><p>Assign arbitrary, negative taxIDs to taxa under this taxon</p></li>
</ul>
</div></blockquote>
</section>
<section id="input-files">
<h3>Input files<a class="headerlink" href="#input-files" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>lineages (<a class="reference internal" href="io.html#lineage-tsv-without-taxid-io"><span class="std std-ref">lineage tsv without taxID</span></a>; output of <a class="reference internal" href="#format-custom-reference"><span class="std std-ref">format_custom.pl</span></a> or <a class="reference internal" href="#format-bold-package-reference"><span class="std std-ref">format_bold_package.pl</span></a>)</p></li>
<li><p>sequences (<a class="reference internal" href="io.html#sequence-tsv-without-taxid-io"><span class="std std-ref">sequence tsv without taxID</span></a>)</p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="parameters-options">
<h3>Parameters/options<a class="headerlink" href="#parameters-options" title="Link to this heading"></a></h3>
<p>NONE</p>
</section>
<section id="algorithm">
<h3>Algorithm<a class="headerlink" href="#algorithm" title="Link to this heading"></a></h3>
<p>For each lineage finds the lowest taxon that matches an already existing <a class="reference internal" href="glossary.html#taxid-glossary"><span class="std std-ref">taxID</span></a> :</p>
<blockquote>
<div><ul class="simple">
<li><p>Finds the smallest taxon that matches an already existing taxID (taking into account all taxIDs in the input taxonomy file)</p></li>
<li><p>Accepts taxID if at least 0.6 of the taxa in the upward input lineages matches the lineage of the taxID (for species level do not count the genus, since it matches necessarily the species name) OR Both the input taxon and taxID have a phylum rank</p></li>
<li><p>If the match between the input lineage and the taxID lineage is &lt;=0.25, go to the next taxlevel</p></li>
<li><p>If the match between the input lineage and the taxID lineage is between 0.25 and 0.6, print lineage to ambiguous file. It should be checked manually.</p></li>
</ul>
</div></blockquote>
<p>The search for an existing taxID, takes into account</p>
<blockquote>
<div><ul class="simple">
<li><p>Existing NCBI or previous arbitrary (negative) taxIDs</p></li>
<li><p>All synonym taxon names in the taxonomy file</p></li>
<li><p>If there are more than one taxIDs for the taxon name, take the one with the highest proportion of taxa matching the upwards lineage and then the one that has the same taxonomic rank</p></li>
</ul>
</div></blockquote>
<p>If there is no taxID for the lowest taxonomic levels in the lineage, assign arbitrary negative taxIDs to them.
New taxIDs are linked as a child to an existing taxID, and the taxonomy file is updated with them.</p>
</section>
<section id="output">
<h3>Output<a class="headerlink" href="#output" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequences_with_taxIDs.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#lineage-tsv-with-taxid-add-taxids-io"><span class="std std-ref">lineages_with_taxIDs.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#ambiguous-lineages-io"><span class="std std-ref">ambiguous_lineages.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#sequence-tsv-without-taxid-io"><span class="std std-ref">ambiguous_sequences.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy_updated.tsv</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="dereplicate-pl">
<span id="dereplicate-reference"></span><h2>dereplicate.pl<a class="headerlink" href="#dereplicate-pl" title="Link to this heading"></a></h2>
<section id="id1">
<h3>Aim<a class="headerlink" href="#id1" title="Link to this heading"></a></h3>
<p>Taxonomically aware dereplication.</p>
</section>
<section id="id2">
<h3>Input files<a class="headerlink" href="#id2" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">tsv sequence tsv with taxID</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
<li><p>out (name of the output dereplicated sequence tsv file)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id3">
<h3>Parameters/options<a class="headerlink" href="#id3" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>vsearch_path (path to vsearch executables if not in the PATH)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id4">
<h3>Algorithm<a class="headerlink" href="#id4" title="Link to this heading"></a></h3>
<p>Compare sequences of the same taxID.
Delete sequences that are substring of another sequence (100% identity on the overlapping region, and one sequence covers the other completely).
If more than 10.000 sequences for the same taxID, first, cluster the sequences with 100% of identity using the cluster_fast algorithm of vsearch, then use the substring search for each cluster.</p>
</section>
<section id="id5">
<h3>Output<a class="headerlink" href="#id5" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">tsv sequence tsv with taxID</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="download-bold-pl">
<span id="download-bold-reference"></span><h2>download_bold.pl<a class="headerlink" href="#download-bold-pl" title="Link to this heading"></a></h2>
<section id="id6">
<h3>Aim<a class="headerlink" href="#id6" title="Link to this heading"></a></h3>
<p>Download BOLD data in tsv format for a list of taxa.</p>
</section>
<section id="id7">
<h3>Input files<a class="headerlink" href="#id7" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#taxon-list-io"><span class="std std-ref">taxon_list</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="id8">
<h3>Parameters/options<a class="headerlink" href="#id8" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>try_download (integer; Try to download files <em>try_download</em> times if some of the downloaded files are incomplete; Default: 3)</p></li>
<li><p>max_record_n (integer; If more than max_record_n records for a taxon, cut up the taxon to subtaxa; Do not cut up input taxa if max_record_n is 0; Default: 0)</p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> Only necessary id max_record_n &gt; 0</p></li>
</ul>
</div></blockquote>
</section>
<section id="id9">
<h3>Algorithm<a class="headerlink" href="#id9" title="Link to this heading"></a></h3>
<p>The script downloads all sequences and lineages for all taxa on the taxon_list from BOLD.</p>
<p>The taxon_list file was constructed manually from taxa on <a class="reference external" href="https://www.boldsystems.org/index.php/TaxBrowser_Home">https://www.boldsystems.org/index.php/TaxBrowser_Home</a>. Each taxon on the list has less than 750.000 specimen records in BOLD.
The taxon_list constructed on 2022-02-24 is available on github (data/bold_taxon_list_2022-02-24.txt). It contains all taxa available in BOLD. This file might need to be updated later.</p>
<p>For constructing a taxon specific taxon_list (e.g. include only taxa of Arthropoda), the
<a class="reference external" href="https://github.com/meglecz/mkCOInr/blob/main/data/bold_taxon_list_2022-02-24_details.tsv">bold_taxon_list_2022-02-24_details.tsv</a>
file is available in the data directory, where the lineage of each taxon and the number of specimen records on are also included.
You can use this file to easily select the taxa of your interest and make a custom taxon_list.</p>
<p>Alternatively, a list of taxa (including large taxa such as Arthropoda) can be given and the input taxa can be cut up automatically to subtaxa of less than <em>max_record_n</em> records each.
This method has the advantage of avoiding the manual construction of the taxon list (as for bold_taxon_list_2022-02-24.txt).
However, the subtaxa produced by the script are based on the NCBI taxonomy, and in case of divergent nomenclature between BOLD and NCBI,
some of the subtaxa can be missed.
For example, if the Chordata phylum is cut up to classes, it will contain the Lepidosauria class. Lepidosauria gives 0 results,
since in BOLD the class field contains Reptilia instead of Lepidosauria, thus missing out BOLD orders like Crocodylia, Rhynchocephalia, Squamata, Testudines.</p>
<p>Download is done using BOLD’s API. First a small stat file is downloaded to access the number of records available for the taxa, then the tsv file is downloaded with sequences and metainfo.
The script checks if the number of downloaded records corresponds to the expected one (based on the stat file).
If there is an error, it removes the file and retries the download try_download times.</p>
<p>If the file exists already, the download is skipped. In this way, if the program stops (for example hitting wall time on a server), it can be simply restarted and the taxa with successful downloads will not be rerun.</p>
<p>NOTE: The download for a long list of taxa can take several days since it is not parallelized.
You can cut up the input list and run each of them on a separate computer and move the output files to the same folder afterwards.</p>
</section>
<section id="id10">
<h3>Output<a class="headerlink" href="#id10" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>json file for each taxon with the total number of records</p></li>
<li><p>tsv file for each taxon with the following columns:</p></li>
</ul>
<p>processid   sampleid        recordID        catalognum      fieldnum        institution_storing     collection_code bin_uri phylum_taxID    phylum_name     class_taxID     class_name      order_taxID     order_name      family_taxID    family_name     subfamily_taxID subfamily_name  genus_taxID     genus_name      species_taxID   species_name    subspecies_taxID        subspecies_name identification_provided_by      identification_method   identification_reference        tax_note        voucher_status  tissue_type     collection_event_id     collectors      collectiondate_start    collectiondate_end      collectiontime  collection_notesite_code        sampling_protocol       lifestage       sex     reproduction    habitat associated_specimens    associated_taxa extrainfo       notes   lat     lon     coord_source    coord_accuracy  elev    depth   elev_accuracy   depth_accuracy  country province_state  region  sector  exactsite       image_ids       image_urls      media_descriptors       captions        copyright_holders       copyright_years copyright_licenses      copyright_institutions  photographers   sequenceID      markercode      genbank_accession       nucleotides     trace_ids       trace_names     trace_links     run_dates       sequencing_centers      directions      seq_primers     marker_codes</p>
</div></blockquote>
</section>
</section>
<section id="download-taxonomy-pl">
<span id="download-taxonomy-reference"></span><h2>download_taxonomy.pl<a class="headerlink" href="#download-taxonomy-pl" title="Link to this heading"></a></h2>
<section id="id11">
<h3>Aim<a class="headerlink" href="#id11" title="Link to this heading"></a></h3>
<p>Download the NCBI taxonomy dmp files (<a class="reference external" href="https://ftp.ncbi.nih.gov/pub/taxonomy/new_taxdump/">https://ftp.ncbi.nih.gov/pub/taxonomy/new_taxdump/</a>) and prepare <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> file.</p>
</section>
<section id="input">
<h3>Input<a class="headerlink" href="#input" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="id12">
<h3>Parameters/options<a class="headerlink" href="#id12" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>skip_download (0/1; if 1, skips download, only prepares taxonomy file; Default: 0)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id13">
<h3>Algorithm<a class="headerlink" href="#id13" title="Link to this heading"></a></h3>
<p>Downloads ncbi taxonomy dump files from <a class="reference external" href="https://ftp.ncbi.nih.gov/pub/taxonomy/new_taxdump/">https://ftp.ncbi.nih.gov/pub/taxonomy/new_taxdump/</a> to the ncbi_tax subdirectory and decompress them.
Prepares a <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> file with all taxIDs in it.</p>
</section>
<section id="id14">
<h3>Output<a class="headerlink" href="#id14" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>ncbi_tax subdirectory with dmp files</p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="format-bold-pl">
<span id="format-bold-reference"></span><h2>format_bold.pl<a class="headerlink" href="#format-bold-pl" title="Link to this heading"></a></h2>
<section id="id15">
<h3>Aim<a class="headerlink" href="#id15" title="Link to this heading"></a></h3>
<p>Prepare a single sequence file and a lineage file from all tsv files downloaded from BOLD.
Clean and orient sequences.</p>
</section>
<section id="id16">
<h3>Input files<a class="headerlink" href="#id16" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>download_dir (name of the folder containing the downloaded BOLD tsv files)</p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="id17">
<h3>Parameters/options<a class="headerlink" href="#id17" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>marker_list (List of markers to be selected; Default:  ‘COI-5P COI-3P’)</p></li>
<li><p>check_name (0/1; If 1 keeps only taxa with valid Latin name format: Default: 1)</p></li>
<li><p>max_n (positive integer; eliminates sequences with max_n or more consecutive Ns; Default:5)</p></li>
<li><p>min_length (positive integer; minimum length of the cleaned sequence; Default:100)</p></li>
<li><p>max_length (positive integer; maximum length of the cleaned sequence; Default:2000)</p></li>
<li><p>check_orientation (0/1; if 1, checks the orientation of the sequences; Default: 0)</p></li>
<li><p>blast_path (Optional; Path to the BLAST executables if they is not in your PATH)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id18">
<h3>Algorithm<a class="headerlink" href="#id18" title="Link to this heading"></a></h3>
<p>Clean downloaded files and pool information to lineage and sequence files</p>
<blockquote>
<div><ul class="simple">
<li><p>Eliminate partial lines (mostly errors in the database)</p></li>
<li><p>Select sequences for a given marker list</p></li>
<li><dl class="simple">
<dt>Clean sequences</dt><dd><ul>
<li><p>Correct sequence IDs</p></li>
<li><p>Gaps deleted</p></li>
<li><p>Non-TCGA changed to N</p></li>
<li><p>External Ns deleted</p></li>
<li><p>Sequences with more than max_n consecutive Ns are deleted</p></li>
<li><p>Keep sequences with length in a min_length and max_length range</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Clean lineages</dt><dd><ul>
<li><p>If check_name, keep only names matching a correct Latin name format (only letters, spaces and -, correct capitalization)</p></li>
<li><p>Pool identical lineages into one line with the list of valid sequence IDs in the last field</p></li>
<li><p>Eliminate lines with environmental and metagenomic samples</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Orient sequence (optional)</dt><dd><ul>
<li><p>Count the TAA, TAG STOP codons in each reading frame</p></li>
<li><p>Choose the orientation where there is no STOP codon</p></li>
<li><p>If there is a STOP codon in all frames OR there are frames without STOP codon both in stand + and -, class it as ambiguous</p></li>
<li><p>Make a small “reference” db form randomly sampled oriented sequences</p></li>
<li><p>Blast ambiguous sequences to check orientation</p></li>
<li><p>Write sequences without hit to the bold_ambiguous_orientation.fas</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</section>
<section id="id19">
<h3>Output<a class="headerlink" href="#id19" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-without-taxid-io"><span class="std std-ref">bold_sequences.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#lineage-tsv-without-taxid-io"><span class="std std-ref">bold_lineages.tsv</span></a></p></li>
<li><p>bold_partial_lines.tsv (lines in the input tsv files that did not have sequences)</p></li>
<li><p>bold_ambiguous_orientation.fas (sequences that could not be oriented in check_orientation option)</p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="format-bold-package-pl">
<span id="format-bold-package-reference"></span><h2>format_bold_package.pl<a class="headerlink" href="#format-bold-package-pl" title="Link to this heading"></a></h2>
<section id="id20">
<h3>Aim<a class="headerlink" href="#id20" title="Link to this heading"></a></h3>
<p>Prepare a single sequence file and a lineage file from a BOLD data package file downloaded from <cite>BOLD &lt;https://www.boldsystems.org/index.php/datapackages&gt;_</cite>
Clean and orient sequences.</p>
</section>
<section id="id21">
<h3>Input files<a class="headerlink" href="#id21" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><span class="xref std std-ref">bold_data</span> (name of the data package downloaded from BOLD; tsv format)</p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="id22">
<h3>Parameters/options<a class="headerlink" href="#id22" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>marker_list (List of markers to be selected; Default:  ‘COI-5P COI-3P’)</p></li>
<li><p>check_name (0/1; If 1 keeps only taxa with valid Latin name format: Default: 1)</p></li>
<li><p>max_n (positive integer; eliminates sequences with max_n or more consecutive Ns; Default:5)</p></li>
<li><p>min_length (positive integer; minimum length of the cleaned sequence; Default:100)</p></li>
<li><p>max_length (positive integer; maximum length of the cleaned sequence; Default:2000)</p></li>
<li><p>check_orientation (0/1; if 1, checks the orientation of the sequences; Default: 0)</p></li>
<li><p>blast_path (Optional; Path to the BLAST executables if they is not in your PATH)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id23">
<h3>Algorithm<a class="headerlink" href="#id23" title="Link to this heading"></a></h3>
<p>Select and clean sequences and pool information to lineage and sequence files</p>
<blockquote>
<div><ul class="simple">
<li><p>Eliminate partial lines (mostly errors in the database)</p></li>
<li><p>Select sequences for a given marker list</p></li>
<li><dl class="simple">
<dt>Clean sequences</dt><dd><ul>
<li><p>Correct sequence IDs</p></li>
<li><p>Gaps deleted</p></li>
<li><p>Non-TCGA changed to N</p></li>
<li><p>External Ns deleted</p></li>
<li><p>Sequences with more than max_n consecutive Ns are deleted</p></li>
<li><p>Keep sequences with length in a min_length and max_length range</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Clean lineages</dt><dd><ul>
<li><p>If check_name, keep only names matching a correct Latin name format (only letters, spaces and -, correct capitalization)</p></li>
<li><p>Pool identical lineages into one line with the list of valid sequence IDs in the last field</p></li>
<li><p>Eliminate lines with environmental and metagenomic samples</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Orient sequence (optional)</dt><dd><ul>
<li><p>Count the TAA, TAG STOP codons in each reading frame</p></li>
<li><p>Choose the orientation where there is no STOP codon</p></li>
<li><p>If there is a STOP codon in all frames OR there are frames without STOP codon both in stand + and -, class it as ambiguous</p></li>
<li><p>Make a small “reference” db form randomly sampled oriented sequences</p></li>
<li><p>Blast ambiguous sequences to check orientation</p></li>
<li><p>Write sequences without hit to the bold_ambiguous_orientation.fas</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</section>
<section id="id24">
<h3>Output<a class="headerlink" href="#id24" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-without-taxid-io"><span class="std std-ref">bold_sequences.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#lineage-tsv-without-taxid-io"><span class="std std-ref">bold_lineages.tsv</span></a></p></li>
<li><p>bold_partial_lines.tsv (lines in the input tsv files that did not have sequences)</p></li>
<li><p>bold_ambiguous_orientation.fas (sequences that could not be oriented in check_orientation option)</p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="format-custom-pl">
<span id="format-custom-reference"></span><h2>format_custom.pl<a class="headerlink" href="#format-custom-pl" title="Link to this heading"></a></h2>
<section id="id25">
<h3>Aim<a class="headerlink" href="#id25" title="Link to this heading"></a></h3>
<p>Make a lineage file for each input taxon name.
Prepare input for add_taxid.pl</p>
<p>The output lineage file should be checked manually</p>
<blockquote>
<div><ul class="simple">
<li><p>To see if the suggested lineages are plausible</p></li>
<li><p>To select the correct lineage if there is more than one (1 in homonymy column) for the same taxon name</p></li>
</ul>
</div></blockquote>
<p>Before the next step (add_taxids.pl)</p>
<blockquote>
<div><ul class="simple">
<li><p>Correct/delete/complete lines if lineage in not correct. Try to use taxon names compatible with NCBI taxonomy (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/taxonomy/">https://www.ncbi.nlm.nih.gov/taxonomy/</a> )</p></li>
<li><p>Delete the homonymy column</p></li>
</ul>
</div></blockquote>
</section>
<section id="id26">
<h3>Input<a class="headerlink" href="#id26" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#custom-sequences-tsv-io"><span class="std std-ref">custom sequences</span></a></p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="id27">
<h3>Parameters/options<a class="headerlink" href="#id27" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>max_n (positive integer; eliminate sequences with max_n or more consecutive Ns; Default: 5)</p></li>
<li><p>min_length (positive integer; minimum length of the cleaned sequence; Default: 100)</p></li>
<li><p>max_length (positive integer; maximum length of the cleaned sequence; Default: 2000 )</p></li>
<li><p>check_seqid_format (0/1; if 1 check if seqID resemble to bold and ncbi formats, print out warnings, if yes; Default: 1)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id28">
<h3>Algorithm<a class="headerlink" href="#id28" title="Link to this heading"></a></h3>
<p>Match names to all taxon names in <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> (including synonyms)</p>
<blockquote>
<div><ul class="simple">
<li><p>Write a lineage to all taxon names where name matches to an existing name in taxonomy.tsv</p></li>
<li><p>If homonymy, create a lineage for each homonym, and write 1 to the homonymy column</p></li>
<li><p>If taxon name corresponds to a Latin name format (Genus species) and species name is not known, get the lineage for the genus.</p></li>
</ul>
</div></blockquote>
<p>If the check_seqid_format option is activated</p>
<blockquote>
<div><ul class="simple">
<li><p>If some of the sequence IDs are not unique, list the duplicates IDs and exit</p></li>
<li><p>If sequence ID format is similar to accessions used in BOLD and NCBI/EMBL/DDBJ, list IDs but continue</p></li>
<li><p>A fairly safe format is xxx_xxx####, where x is a letter, # is a digit</p></li>
</ul>
</div></blockquote>
<p>Clean sequences</p>
<blockquote>
<div><ul class="simple">
<li><p>gaps deleted</p></li>
<li><p>non-TCGA changed to N</p></li>
<li><p>external Ns deleted</p></li>
<li><p>sequences with more than max_n consecutive Ns are deleted</p></li>
</ul>
</div></blockquote>
</section>
<section id="id29">
<h3>Output<a class="headerlink" href="#id29" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#custom-lineages-tsv-io"><span class="std std-ref">custom_lineages.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#sequence-tsv-without-taxid-io"><span class="std std-ref">custom_sequences.tsv</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="format-db-pl">
<span id="format-db-reference"></span><h2>format_db.pl<a class="headerlink" href="#format-db-pl" title="Link to this heading"></a></h2>
<section id="id30">
<h3>Aim<a class="headerlink" href="#id30" title="Link to this heading"></a></h3>
<p>Make a database in blast, rdp, qiime or full tsv format from the <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv</span></a> and <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> files</p>
</section>
<section id="id31">
<h3>Input<a class="headerlink" href="#id31" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a></p></li>
<li><p>outfmt (rdp/blast/qiime/full/vtam/sintax; choose the format of the database)</p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
<li><p>out (string for naming the output files)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id32">
<h3>Parameters/options<a class="headerlink" href="#id32" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>blast_path (Optional; path to the blast executables if it is not in your PATH)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id33">
<h3>Algorithm<a class="headerlink" href="#id33" title="Link to this heading"></a></h3>
<p>BLAST db, VTAM</p>
<blockquote>
<div><ul class="simple">
<li><p>Prepare a fasta file with sequences and the taxIDs (seqID, taxID)</p></li>
<li><p>Run the <em>makeblastdb</em> command of blast to make indexed files ready to be used as a blast database</p></li>
<li><p>for VTAM format, prepare taxonomy file as well as the BLAST database. They can be used directly in VTAM.</p></li>
</ul>
</div></blockquote>
<p>RDP, QIIME, sintax and FULL</p>
<blockquote>
<div><ul class="simple">
<li><p>Prepare a ranked lineage for each taxID</p></li>
<li><p>Taxon names are concatenated with taxID to avoid homonymy</p></li>
<li><p>Missing taxonomic levels are completed by using the name of a higher-level taxon concatenated with the taxonomic ranks</p></li>
<li><p>Prepare a trainseq fasta and a taxon file for <a class="reference internal" href="io.html#rdp-io"><span class="std std-ref">rdp</span></a> and <a class="reference internal" href="io.html#qiime-io"><span class="std std-ref">qiime</span></a></p></li>
<li><p>Prepare a single tsv file for <a class="reference internal" href="io.html#full-tsv-io"><span class="std std-ref">full</span></a></p></li>
<li><p>Prepare a single fasta file with lineages incuded in the definition lines for <a class="reference internal" href="io.html#sintax-io"><span class="std std-ref">sintax</span></a></p></li>
</ul>
</div></blockquote>
<p>The trainseq fasta and the taxon files can be used by the <em>train</em> command of rdp_classifier or <em>feature-classifier</em> of QIIME to train the dataset before classification.</p>
<p>The sintax fasta file is ready to be used as a database for SINTAX.</p>
<p>The full tsv format is an easy to parse tsv file with <a class="reference internal" href="glossary.html#ranked-lineage-glossary"><span class="std std-ref">ranked lineage</span></a> and <a class="reference internal" href="glossary.html#taxid-glossary"><span class="std std-ref">taxID</span></a> for each sequence.</p>
</section>
<section id="id34">
<h3>Output<a class="headerlink" href="#id34" title="Link to this heading"></a></h3>
<p>blast option</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#blast-database-files-io"><span class="std std-ref">Indexed files</span></a> ready to be used as a BLAST database</p></li>
</ul>
</div></blockquote>
<p>vtam option</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#blast-database-files-io"><span class="std std-ref">Indexed files</span></a> ready to be used as a BLAST database</p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> adapted to VTAM</p></li>
</ul>
</div></blockquote>
<p>rdp option</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#rdp-trainseq-fasta-io"><span class="std std-ref">RDP trainseq fasta</span></a></p></li>
<li><p><a class="reference internal" href="io.html#rdp-taxon-file-io"><span class="std std-ref">RDP taxon file</span></a></p></li>
</ul>
</div></blockquote>
<p>quiime option</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#qiime-trainseq-fasta-io"><span class="std std-ref">QIIME trainseq fasta</span></a></p></li>
<li><p><a class="reference internal" href="io.html#qiime-taxon-file-io"><span class="std std-ref">QIIME taxon file</span></a></p></li>
</ul>
</div></blockquote>
<p>sintax option</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sintax-fasta-io"><span class="std std-ref">SINTAX fasta</span></a></p></li>
</ul>
</div></blockquote>
<p>full option</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#full-tsv-io"><span class="std std-ref">tsv file</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="format-rdp-pl">
<span id="format-rdp-reference"></span><h2>format_rdp.pl<a class="headerlink" href="#format-rdp-pl" title="Link to this heading"></a></h2>
<section id="id35">
<h3>Aim<a class="headerlink" href="#id35" title="Link to this heading"></a></h3>
<p>Format the RDP training dataset to <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv file with taxIDs</span></a> and <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a></p>
</section>
<section id="id36">
<h3>Input files<a class="headerlink" href="#id36" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#rdp-classifier-trainset-fasta-io"><span class="std std-ref">RDP classifier trainset fasta file</span></a></p></li>
<li><p><a class="reference internal" href="io.html#rdp-classifier-taxid-file-io"><span class="std std-ref">RDP classifier taxid file</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="id37">
<h3>Parameters/options<a class="headerlink" href="#id37" title="Link to this heading"></a></h3>
<blockquote>
<div><p>NA</p>
</div></blockquote>
</section>
<section id="id38">
<h3>Algorithm<a class="headerlink" href="#id38" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>Reformat the rdp taxonomy file to <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> format</p></li>
<li><p>Make <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv file with taxIDs</span></a> from the fasta file</p></li>
</ul>
</div></blockquote>
</section>
<section id="id39">
<h3>Output<a class="headerlink" href="#id39" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence.tsv with taxIDs</span></a></p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="format-ncbi-pl">
<span id="format-ncbi-reference"></span><h2>format_ncbi.pl<a class="headerlink" href="#format-ncbi-pl" title="Link to this heading"></a></h2>
<section id="id40">
<h3>Aim<a class="headerlink" href="#id40" title="Link to this heading"></a></h3>
<p>Format the CDS and taxID files to a <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv file with taxIDs</span></a>.
Select and clean sequences.</p>
</section>
<section id="id41">
<h3>Input files<a class="headerlink" href="#id41" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>cds (CDS fasta file; output of nsdpy)</p></li>
<li><p>taxids (tsv file with the seqID and taxID columns; output of nsdpy)</p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="id42">
<h3>Parameters/options<a class="headerlink" href="#id42" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>check_name (0/1; If one, keep only taxa with valid Latin name fomat: Default: 1)</p></li>
<li><p>max_n (positive integer; eliminate sequences with max_n or more consecutive Ns; Default: 5)</p></li>
<li><p>min_length (positive integer; minimum length of the cleaned sequence; Default: 100)</p></li>
<li><p>max_length (positive integer; maximum length of the cleaned sequence; Default: 2000)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id43">
<h3>Algorithm<a class="headerlink" href="#id43" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><dl class="simple">
<dt>Select sequences</dt><dd><ul>
<li><p>Check if gene names and protein names correspond to COI</p></li>
<li><p>Eliminate genes if they have introns</p></li>
<li><p>Can have more than one COI gene in the same mitochondrion</p></li>
<li><p>Accept only if valid taxID, replace old non-valid taxIDs by up-to-date taxIDs</p></li>
<li><p>Eliminate sequences from environmental and metagenomic samples</p></li>
<li><p>If check_name is activated, take the taxID of the smallest taxon with a valid Latin name, otherwise keep the original taxID.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Clean sequences</dt><dd><ul>
<li><p>Upper case</p></li>
<li><p>Replace non-ATCG by N</p></li>
<li><p>Delete gaps and external Ns</p></li>
<li><p>Delete sequence if more than max_n consecutive Ns</p></li>
<li><p>Keep sequences if length is between min_length and max_length</p></li>
<li><p>Sequences are already in a correct orientation in the input file, since that are coming from CDS files</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</section>
<section id="id44">
<h3>Output<a class="headerlink" href="#id44" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">ncbi_sequences.tsv</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="get-subtaxa-pl">
<span id="get-subtaxa-reference"></span><h2>get_subtaxa.pl<a class="headerlink" href="#get-subtaxa-pl" title="Link to this heading"></a></h2>
<section id="id45">
<h3>Aim<a class="headerlink" href="#id45" title="Link to this heading"></a></h3>
<p>List subtaxa of the input taxon at the next major taxonomic rank (e.g. list all orders of the input class)</p>
</section>
<section id="id46">
<h3>Input files<a class="headerlink" href="#id46" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>taxon (can be a taxon name or taxID)</p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
</ul>
</div></blockquote>
</section>
<section id="id47">
<h3>Parameters/options<a class="headerlink" href="#id47" title="Link to this heading"></a></h3>
<p>NONE</p>
</section>
<section id="id48">
<h3>Algorithm<a class="headerlink" href="#id48" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>If taxon is a taxon name, get all taxIDs that correspond to this name (e.g. 1266065 and 50622 for Plecoptera)</p></li>
<li><p>Determine the next lowest major taxonomic rank (phylum, class, order, family, genus, species) for each taxID (e.g. if taxId is an order or suborder or superfamily, the next major tax rank is family)</p></li>
<li><p>List subtaxa of each taxID of this taxonomic rank.</p></li>
</ul>
</section>
<section id="id49">
<h3>Output<a class="headerlink" href="#id49" title="Link to this heading"></a></h3>
<ul class="simple">
<li><dl class="simple">
<dt>tsv with the following columns</dt><dd><ul>
<li><p>taxon</p></li>
<li><p>taxID</p></li>
<li><p>subtaxon</p></li>
<li><p>taxID (of the subtaxa)</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</section>
</section>
<section id="pool-and-dereplicate-pl">
<span id="pool-and-dereplicate-reference"></span><h2>pool_and_dereplicate.pl<a class="headerlink" href="#pool-and-dereplicate-pl" title="Link to this heading"></a></h2>
<section id="id50">
<h3>Aim<a class="headerlink" href="#id50" title="Link to this heading"></a></h3>
<p>Pool 2 dereplicated sequence tsv files and do a taxonomically-aware dereplication for taxIDs present in both input files</p>
</section>
<section id="id51">
<h3>Input files<a class="headerlink" href="#id51" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">tsv1</span></a></p></li>
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">tsv2</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
<li><p>out (name of the output dereplicated sequence tsv file)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id52">
<h3>Parameters/options<a class="headerlink" href="#id52" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>vsearch_path (path to vsearch executables if not in the PATH)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id53">
<h3>Algorithm<a class="headerlink" href="#id53" title="Link to this heading"></a></h3>
<p>Pool sequences from two input tsv files.
The dereplication is done only for taxID shared by the two input files, since they have been dereplicated individually.
The algorithm of dereplication is identical to the one used in <a class="reference internal" href="#dereplicate-reference"><span class="std std-ref">dereplicate.pl</span></a></p>
</section>
<section id="id54">
<h3>Output<a class="headerlink" href="#id54" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv with taxIDs</span></a></p></li>
</ul>
<p>reduce_metadata.pl
.. _reduce_metadata_reference:</p>
</section>
</section>
<section id="reduce-metadata-pl">
<h2>reduce_metadata.pl<a class="headerlink" href="#reduce-metadata-pl" title="Link to this heading"></a></h2>
<section id="id55">
<h3>Aim<a class="headerlink" href="#id55" title="Link to this heading"></a></h3>
<p>Select metadata of BOLD sequences that appear in the final file of COInr.tsv; The output is used to be able to trace back BOLD sequences to their autors and get their metadata.</p>
</section>
<section id="id56">
<h3>Input files<a class="headerlink" href="#id56" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#bold-data-package-io"><span class="std std-ref">input_metadata</span></a></p></li>
<li><p><a class="reference internal" href="io.html#bold-data-package-io"><span class="std std-ref">coinr</span></a></p></li>
<li><p>out (name of the output tsv file)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id57">
<h3>Algorithm<a class="headerlink" href="#id57" title="Link to this heading"></a></h3>
<p>Select lines from the BOLD data package, that reamin in the COInr database.</p>
</section>
<section id="id58">
<h3>Output<a class="headerlink" href="#id58" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="io.html#bold-data-package-io"><span class="std std-ref">sequence tsv with BOLD metadata</span></a></p></li>
</ul>
</section>
</section>
<section id="select-region-pl">
<span id="select-region-reference"></span><h2>select_region.pl<a class="headerlink" href="#select-region-pl" title="Link to this heading"></a></h2>
<section id="id59">
<h3>Aim<a class="headerlink" href="#id59" title="Link to this heading"></a></h3>
<p>Select a target region from input sequences.
As an input, either a primer pair should be given to identify the target region in some sequences by e-pcr, or a fasta file containing taxonomically diverse sequences limited to the target region.
The sequences are then aligned to the target sequences and trimmed according to the alignment</p>
</section>
<section id="id60">
<h3>Input files<a class="headerlink" href="#id60" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv with taxIDs</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
<li><p>bait_fas (A small phylogenetically diverse fasta file with sequences already trimmed to the target region; Optional; Can be produced by e-pcr included in the script.)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id61">
<h3>Parameters/options<a class="headerlink" href="#id61" title="Link to this heading"></a></h3>
<p><em>E-pcr related parameters</em></p>
<blockquote>
<div><ul class="simple">
<li><p>e_pcr  (0/1; if 1, identify the target region of the sequences by e-pcr in the first step)</p></li>
<li><p>fw (optional; if e_pcr is done, the sequence of the forward primer that amplifies the target region)</p></li>
<li><p>rv (optional; if e_pcr is done, the sequence of the reverse primer that amplifies the target region)</p></li>
<li><p>trim_error (real [0-1], the proportion of mismatches allowed between the primer and the sequence during the e_pcr; Default: 0.3)</p></li>
<li><p>min_overlap (the minimum overlap between primer and the sequence during e-pcr; Default: 20)</p></li>
<li><p>min_amplicon_length (The minimum length of the amplicon after primer trimming; Default: 100)</p></li>
<li><p>max_amplicon_length (The minimum length of the amplicon after primer trimming; Default: 2000)</p></li>
<li><p>cutadapt_path (Optional; Path to cutadapt if it is not in your PATH)</p></li>
</ul>
</div></blockquote>
<p><em>usearch_global related parameters</em> for trimming sequences according to the alignments</p>
<blockquote>
<div><ul class="simple">
<li><p>tcov (minimum coverage of the target sequence in <em>usearch_global</em> hits; Default: 0.5)</p></li>
<li><p>identity (minimum identity between the sequence and the target in <em>usearch_global</em> hits; Default: 0.7)</p></li>
<li><p>vsearch_path (Optional; path to the vsearch if it is not in your PATH)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id62">
<h3>Algorithm<a class="headerlink" href="#id62" title="Link to this heading"></a></h3>
<p>A fasta file is prepared from the input tsv sequence file.</p>
<p>Sequences are aligned to a small pool of target sequences already limited to the target region (bait_fas).
The alignment is done by the <em>usearch_global</em> command of vsearch which makes global alignments (unlike BLAST).</p>
<p>The best hit is used for each sequence to orient and trim them to the target region.
Only hits with a minimum target coverage (tcov) and percentage of identity (identity) are used.</p>
<p>The bait_fas file can be either previously prepared by the users and given as an input,
or be produced by e-pcr using the e-pcr related parameters.</p>
<p>To reduce runtime, sequences in the bait_fas are clustered by -cluster_fast algorithm of vsearch
and the centroids are used as a target file for the <em>usearch_global</em></p>
</section>
<section id="id63">
<h3>Output<a class="headerlink" href="#id63" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>cutadapt_trimmed.fas (if e_pcr; fasta file with sequences recognized and trimmed by E-pcr; equivalent of bait_fas )</p></li>
<li><p>target_centroids.fas (fasta file of centroids of the clustering of bait_fas or cutadapt_trimmed.fas)</p></li>
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">trimmed.tsv</span></a> (sequence tsv files with taxIDs trimmed to the target region)</p></li>
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">untrimmed.tsv</span></a> (sequence tsv files with taxIDs where the target region is not found)</p></li>
</ul>
</div></blockquote>
</section>
</section>
<section id="select-taxa-pl">
<span id="select-taxa-reference"></span><h2>select_taxa.pl<a class="headerlink" href="#select-taxa-pl" title="Link to this heading"></a></h2>
<section id="id64">
<h3>Aim<a class="headerlink" href="#id64" title="Link to this heading"></a></h3>
<p>Select or eliminate  sequences that belong to taxa on a taxon list.
Select sequences with a minimum taxonomic resolution (e.g. assigned at least to genus).</p>
</section>
<section id="id65">
<h3>Input files<a class="headerlink" href="#id65" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv with taxIDs</span></a></p></li>
<li><p><a class="reference internal" href="io.html#taxon-list-io"><span class="std std-ref">taxon_list</span></a></p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy</span></a></p></li>
<li><p><a class="reference internal" href="io.html#outdir-io"><span class="std std-ref">outdir</span></a></p></li>
<li><p>out (name of the output dereplicated sequence tsv file)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id66">
<h3>Parameters/options<a class="headerlink" href="#id66" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p>negative_list (1/0; if 1, keeps all taxa except the ones on the taxon list; Default: 0)</p></li>
<li><p>min_taxlevel (species/genus/family/order/class/phylum/kingdom/root; Default: root)</p></li>
</ul>
</div></blockquote>
</section>
<section id="id67">
<h3>Algorithm<a class="headerlink" href="#id67" title="Link to this heading"></a></h3>
<p>Select sequences that belong to the taxa on a taxon list if <em>negative_list==0</em> (default). Select sequences that DO NOT belong to the taxa on in a taxon list, if <em>negative_list==1</em>.</p>
<p>Keep only sequences that are assigned to at least to <em>min_taxlevel</em> rank.</p>
<p>If taxIDs are not given in the taxon_list file the script uses all taxIDs that matches the taxon name.</p>
<p>A lineage file is written for each taxon in taxon_list and the corresponding taxIDs.</p>
<blockquote>
<div><ul class="simple">
<li><p>It should be checked manually if lineages are coherent with the target taxa</p></li>
<li><p>Homonymy column indicates if there are more than one taxID for a taxon</p></li>
<li><p>If there are incoherent lineages, make a new taxon_list file based on the lineage file including taxon names and taxIDs and rerun the script with the new taxon_list.</p></li>
</ul>
</div></blockquote>
</section>
<section id="id68">
<h3>Output<a class="headerlink" href="#id68" title="Link to this heading"></a></h3>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv with taxIDs</span></a></p></li>
<li><p><a class="reference internal" href="io.html#lineage-tsv-with-taxid-select-taxa-io"><span class="std std-ref">lineage file with taxIDs</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorial.html" class="btn btn-neutral float-left" title="Tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="io.html" class="btn btn-neutral float-right" title="Input/Output" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Emese Meglécz.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>