

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial &mdash; mkCOInr  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9a2dae69"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Description" href="reference.html" />
    <link rel="prev" title="Run mkCOInr" href="run.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            mkCOInr
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview of mkCOInr</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="run.html">Run mkCOInr</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#customize-database">Customize database</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#download-and-untar-coinr">Download and untar COInr</a></li>
<li class="toctree-l3"><a class="reference internal" href="#add-custom-sequences-to-a-database">Add custom sequences to a database</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#format-custom-files">Format custom files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#add-taxids-to-custom-sequences">Add taxIDs to custom sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#dereplicate-custom-sequences">Dereplicate custom sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#pool-and-dereplicate-datasets">Pool and dereplicate datasets</a></li>
<li class="toctree-l4"><a class="reference internal" href="#custom-database">Custom database</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#select-sequences-from-existing-database">Select sequences from existing database</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#select-sequences-for-a-list-of-taxa-with-a-minimum-taxonomic-rank">Select sequences for a list of taxa with a minimum taxonomic rank</a></li>
<li class="toctree-l4"><a class="reference internal" href="#excluding-sequences-of-a-taxon-list">Excluding  sequences of a taxon list</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#select-region">Select region</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#select-region-using-the-e-pcr-option">Select region using the e_pcr option</a></li>
<li class="toctree-l4"><a class="reference internal" href="#select-region-using-the-bait-fas-option">Select region using the bait_fas option</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#format-database">Format database</a></li>
<li class="toctree-l3"><a class="reference internal" href="#chaining-steps-to-make-a-custom-database">Chaining steps to make a custom database</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#exclude-insecta-and-sequences-with-resolution-lower-than-genus">Exclude Insecta and sequences with resolution lower than genus</a></li>
<li class="toctree-l4"><a class="reference internal" href="#keep-eukaryota">Keep Eukaryota</a></li>
<li class="toctree-l4"><a class="reference internal" href="#add-custom-sequences">Add custom sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#keep-only-sequences-with-genus-or-higher-resolution">Keep only sequences with genus or higher resolution</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trim-to-leray-region">Trim to Leray region</a></li>
<li class="toctree-l4"><a class="reference internal" href="#format-for-rdp-classifier">Format for RDP_classifier</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#create-coinr-from-bold-and-ncbi">Create COInr from BOLD and NCBI</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#download-ncbi-taxonomy">Download NCBI taxonomy</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ncbi-sequences">NCBI sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#download-ncbi-sequences">Download NCBI sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#format-ncbi-sequences">Format NCBI sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#dereplicate-ncbi-sequences">Dereplicate NCBI sequences</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#bold-sequences">BOLD sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#download-bold-sequences">Download BOLD sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#format-bold-sequences">Format BOLD sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#add-taxids-to-bold-sequences">Add taxIDs to BOLD sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#dereplicate-bold-sequences">Dereplicate BOLD sequences</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#pool-and-dereplicate-tutorial">Pool and dereplicate datasets</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="reference.html">Description</a></li>
<li class="toctree-l1"><a class="reference internal" href="io.html">Input/Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="reflist.html">List of References</a></li>
<li class="toctree-l1"><a class="reference internal" href="cite.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">mkCOInr</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/content/tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tutorial">
<span id="tutorial-tutorial"></span><h1>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading"></a></h1>
<p>After installing mkCOInr you have a file system like this:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkCOInr
├──<span class="w"> </span>data
│<span class="w">   </span>├──<span class="w"> </span>bold_taxon_list_2022-02-24.txt
│<span class="w">   </span>├──<span class="w"> </span>example
│<span class="w">   </span>│<span class="w">   </span>├──<span class="w"> </span>custom_lineages_verified.tsv
│<span class="w">   </span>│<span class="w">   </span>├──<span class="w"> </span>my_sequences.tsv
│<span class="w">   </span>│<span class="w">   </span>├──<span class="w"> </span>taxon_list_eukaryota.tsv
│<span class="w">   </span>│<span class="w">   </span>├──<span class="w"> </span>taxon_list_insecta.tsv
│<span class="w">   </span>│<span class="w">   </span>└──<span class="w"> </span>taxon_list.tsv
│<span class="w">   </span>└──<span class="w"> </span>one_seq_per_order_658.fas
├──<span class="w"> </span>doc
...<span class="w"> </span><span class="o">(</span>abbreviated<span class="o">)</span>
└──<span class="w"> </span>scripts
<span class="w">        </span>├──<span class="w"> </span>add_taxids.pl
<span class="w">        </span>├──<span class="w"> </span>dereplicate.pl
<span class="w">        </span>├──<span class="w"> </span>download_bold.pl
<span class="w">        </span>├──<span class="w"> </span>download_taxonomy.pl
<span class="w">        </span>├──<span class="w"> </span>format_bold.pl
<span class="w">        </span>├──<span class="w"> </span>format_custom.pl
<span class="w">        </span>├──<span class="w"> </span>format_db.pl
<span class="w">        </span>├──<span class="w"> </span>format_ncbi.pl
<span class="w">        </span>├──<span class="w"> </span>get_subtaxa.pl
<span class="w">        </span>├──<span class="w"> </span>mkdb.pm
<span class="w">        </span>├──<span class="w"> </span>pool_and_dereplicate.pl
<span class="w">        </span>├──<span class="w"> </span>select_region.pl
<span class="w">        </span>└──<span class="w"> </span>select_taxa.pl
</pre></div>
</div>
<section id="customize-database">
<span id="customize-tutorial"></span><h2>Customize database<a class="headerlink" href="#customize-database" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>In the first part of the tutorial, I will start from the COInr database in each major step to illustrate</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#add-custom-sequences-tutorial"><span class="std std-ref">How to include custom sequences</span></a></p></li>
<li><p><a class="reference internal" href="#select-sequences-custom-tutorial"><span class="std std-ref">How to select or eliminate sequences of a list of taxa or a minimum resolution</span></a></p></li>
<li><p><a class="reference internal" href="#select-region-custom-tutorial"><span class="std std-ref">How to select a target region</span></a></p></li>
<li><p><a class="reference internal" href="#format-db-custom-tutorial"><span class="std std-ref">How to format a dataset to different database formats</span></a></p></li>
</ul>
</dd>
</dl>
<p>These steps can be executed independently.</p>
<p>The last example shows how to <a class="reference internal" href="#chained-custom-tutorial"><span class="std std-ref">create a pipeline</span></a> by combining different commands.</p>
<p>The creation of the COInr database is explained in the <a class="reference internal" href="#create-coinr-tutorial"><span class="std std-ref">Create COInr from BOLD and NCBI section</span></a>.
You can download this database from <a class="reference external" href="https://doi.org/10.5281/zenodo.6555985">Zenodo</a> and customize it to your needs.</p>
<section id="download-and-untar-coinr">
<span id="download-coinr-tutorial"></span><h3>Download and untar COInr<a class="headerlink" href="#download-and-untar-coinr" title="Link to this heading"></a></h3>
<p>You will need to change the date in the filename, and get the up-to-date link from zenodo for later releases.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>mkCOInr
wget<span class="w"> </span>https://zenodo.org/record/6555985/files/COInr_2022_05_06.tar.gz
tar<span class="w"> </span>-zxvf<span class="w"> </span>COInr_2022_05_06.tar.gz
rm<span class="w"> </span>COInr_2022_05_06.tar.gz
</pre></div>
</div>
<p>For shortening the paths in this tutorial, rename COInr_2022_05_06 directory to COInr.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mv<span class="w"> </span>COInr_2022_05_06<span class="w"> </span>COInr
</pre></div>
</div>
<p>This gives the following file structure</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkCOInr
├──<span class="w"> </span>COInr
│<span class="w">   </span>├──<span class="w"> </span>COInr.tsv
│<span class="w">   </span>└──<span class="w"> </span>taxonomy.tsv
├──<span class="w"> </span>data
│<span class="w">   </span>├──<span class="w"> </span>bold_taxon_list_2022-02-24.txt
│<span class="w">   </span>├──<span class="w"> </span>example
│<span class="w">   </span>│<span class="w">   </span>├──<span class="w"> </span>custom_lineages_verified.tsv
│<span class="w">   </span>│<span class="w">   </span>├──<span class="w"> </span>my_sequences.tsv
│<span class="w">   </span>│<span class="w">   </span>├──<span class="w"> </span>taxon_list_eukaryota.tsv
│<span class="w">   </span>│<span class="w">   </span>├──<span class="w"> </span>taxon_list_insecta.tsv
│<span class="w">   </span>│<span class="w">   </span>└──<span class="w"> </span>taxon_list.tsv
│<span class="w">   </span>└──<span class="w"> </span>one_seq_per_order_658.fas
...<span class="o">(</span>abbreviated<span class="o">)</span>
└──<span class="w"> </span>scripts
<span class="w">        </span>├──<span class="w"> </span>add_taxids.pl
<span class="w">        </span>├──<span class="w"> </span>dereplicate.pl
<span class="w">        </span>...<span class="o">(</span>abbreviated<span class="o">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>The COInr database is composed of two files</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">COInr.tsv</span></a> contains <a class="reference internal" href="glossary.html#seqid-glossary"><span class="std std-ref">sequenceIDs</span></a>, <a class="reference internal" href="glossary.html#taxid-glossary"><span class="std std-ref">taxIDs</span></a> and sequences</p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> contains all taxIDs and associated information</p></li>
</ul>
</dd>
</dl>
</section>
<section id="add-custom-sequences-to-a-database">
<span id="add-custom-sequences-tutorial"></span><h3>Add custom sequences to a database<a class="headerlink" href="#add-custom-sequences-to-a-database" title="Link to this heading"></a></h3>
<section id="format-custom-files">
<span id="format-custom-tutorial"></span><h4>Format custom files<a class="headerlink" href="#format-custom-files" title="Link to this heading"></a></h4>
<p>The <a class="reference internal" href="io.html#custom-sequences-tsv-io"><span class="std std-ref">input tsv file</span></a> (-custom) contains <a class="reference internal" href="glossary.html#seqid-glossary"><span class="std std-ref">seqIDs</span></a>,
taxon names (can be at any taxonomic level) and sequences (see the example data/example/my_sequences.tsv).
The <a class="reference internal" href="reference.html#format-custom-reference"><span class="std std-ref">format_custom.pl</span></a> script will suggest one or more lineages for each
taxon name based on the existing lineages in <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a>. It will also consider synonyms.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_custom.pl<span class="w"> </span>-custom<span class="w"> </span>data/example/my_sequences.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/custom/1_format
</pre></div>
</div>
<p>The output lineage file (custom_lineages.tsv) looks like this:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>phylum<span class="w">  </span>class<span class="w">   </span>order<span class="w">   </span>family<span class="w">  </span>subfamily<span class="w">       </span>genus<span class="w">   </span>species<span class="w"> </span>homonymy<span class="w">        </span>seqIDs
Mollusca<span class="w">        </span>Bivalvia<span class="w">        </span>Cardiida<span class="w">        </span>Cardiidae<span class="w">               </span>Acanthocardia<span class="w">   </span>Acanthocardia<span class="w"> </span>paucicostata<span class="w">      </span><span class="m">0</span><span class="w">       </span>Seq113<span class="p">;</span>Seq88
NA<span class="w">      </span>NA<span class="w">      </span>NA<span class="w">      </span>NA<span class="w">      </span>NA<span class="w">      </span>NA<span class="w">      </span>Ilia<span class="w"> </span>nucleus<span class="w">    </span><span class="m">0</span><span class="w">       </span>Seq117
Streptophyta<span class="w">    </span>Magnoliopsida<span class="w">   </span>Ericales<span class="w">        </span>Ericaceae<span class="w">               </span>Leucothoe<span class="w">               </span><span class="m">1</span><span class="w">       </span>Seq96
Arthropoda<span class="w">      </span>Malacostraca<span class="w">    </span>Amphipoda<span class="w">       </span>Leucothoidae<span class="w">            </span>Leucothoe<span class="w">               </span><span class="m">1</span><span class="w">       </span>Seq96
Annelida<span class="w">        </span>Polychaeta<span class="w">      </span>Phyllodocida<span class="w">    </span>Polynoidae<span class="w">                              </span><span class="m">0</span><span class="w">       </span>Seq65
</pre></div>
</div>
<p>This output should should be checked manually to see if the lineages are coherent.
If homonymy, choose the correct lineage (e.g. for <em>Leucothoe</em> genus), then delete homonymy column.</p>
<p>If a taxon name is not present in the taxonomy file, the lineage should be completed manually (e.g. <em>Ilia nucleus</em> in the example file).</p>
<p>I created a revised version of the lineage file (data/example/custom_lineages_verified.tsv), which will be used in the next step:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>phylum<span class="w">  </span>class<span class="w">   </span>order<span class="w">   </span>family<span class="w">  </span>subfamily<span class="w">       </span>genus<span class="w">   </span>species<span class="w"> </span>seqIDs
Mollusca<span class="w">        </span>Bivalvia<span class="w">        </span>Cardiida<span class="w">        </span>Cardiidae<span class="w">               </span>Acanthocardia<span class="w">   </span>Acanthocardia<span class="w"> </span>paucicostata<span class="w">      </span>Seq113<span class="p">;</span>Seq88
Arthropoda<span class="w">      </span>Malacostraca<span class="w">    </span>Decapoda<span class="w">        </span>Leucosiidae<span class="w">             </span>Ilia<span class="w">    </span>Ilia<span class="w"> </span>nucleus<span class="w">    </span>Seq117
Arthropoda<span class="w">      </span>Malacostraca<span class="w">    </span>Amphipoda<span class="w">       </span>Leucothoidae<span class="w">            </span>Leucothoe<span class="w">               </span>Seq96
Annelida<span class="w">        </span>Polychaeta<span class="w">      </span>Phyllodocida<span class="w">    </span>Polynoidae<span class="w">                              </span>Seq65
</pre></div>
</div>
<p>See details in description section: <a class="reference internal" href="reference.html#format-custom-reference"><span class="std std-ref">format_custom.pl</span></a> script.</p>
</section>
<section id="add-taxids-to-custom-sequences">
<span id="add-taxids-custom-tutorial"></span><h4>Add taxIDs to custom sequences<a class="headerlink" href="#add-taxids-to-custom-sequences" title="Link to this heading"></a></h4>
<p>The <a class="reference internal" href="reference.html#add-taxids-reference"><span class="std std-ref">add_taxids.pl</span></a> script will</p>
<ul class="simple">
<li><dl class="simple">
<dt>For each lineage in the input file</dt><dd><ul>
<li><p>Find an existing taxID at the lowest possible taxonomic level. <a class="reference internal" href="glossary.html#taxid-glossary"><span class="std std-ref">taxIDs</span></a> can be either from NCBI, or negative taxID already present in <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a>.</p></li>
<li><p>Add new arbitrary (negative) taxIDs to taxa not yet in the taxonomy file</p></li>
<li><p>Link each new taxID to an existing one as a child and include info to the updated taxonomy file</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Make a <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">tsv file with sequences and taxIDs</span></a></p></li>
<li><p>Update the <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a> file</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/add_taxids.pl<span class="w"> </span>-lineages<span class="w"> </span>data/example/custom_lineages_verified.tsv<span class="w"> </span>-sequences<span class="w"> </span>tutorial/custom/1_format/custom_sequences.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/custom/2_add_taxids
</pre></div>
</div>
<p>See details in description section: <a class="reference internal" href="reference.html#add-taxids-reference"><span class="std std-ref">add_taxids.pl</span></a> script.</p>
</section>
<section id="dereplicate-custom-sequences">
<span id="dereplicate-custom-tutorial"></span><h4>Dereplicate custom sequences<a class="headerlink" href="#dereplicate-custom-sequences" title="Link to this heading"></a></h4>
<p>The <a class="reference internal" href="reference.html#dereplicate-reference"><span class="std std-ref">dereplicate.pl</span></a> script will eliminate sequences that are substrings of another sequence of the same <a class="reference internal" href="glossary.html#taxid-glossary"><span class="std std-ref">taxID</span></a>.
Use <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequences_with_taxIDs.tsv</span></a> file (output of the previous script) as the input.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/dereplicate.pl<span class="w"> </span>-tsv<span class="w"> </span>tutorial/custom/2_add_taxids/sequences_with_taxIDs.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/custom/3_dereplicate<span class="w"> </span>-out<span class="w"> </span>custom_dereplicated_sequences.tsv
</pre></div>
</div>
<p>The output file is in the same format as the input tsv file.</p>
<p>See details in description section: <a class="reference internal" href="reference.html#dereplicate-reference"><span class="std std-ref">dereplicate.pl</span></a> script.</p>
</section>
<section id="pool-and-dereplicate-datasets">
<span id="pool-and-dereplicate-custom-tutorial"></span><h4>Pool and dereplicate datasets<a class="headerlink" href="#pool-and-dereplicate-datasets" title="Link to this heading"></a></h4>
<dl class="simple">
<dt>Use two dereplicated <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv files</span></a>:</dt><dd><ul class="simple">
<li><p>COInr.tsv  (pool of BOLD and NCBI, downloaded from Zenodo)</p></li>
<li><p>custom_dereplicated_sequences.tsv (output of the previous script)</p></li>
</ul>
</dd>
</dl>
<p><a class="reference internal" href="reference.html#pool-and-dereplicate-reference"><span class="std std-ref">pool_and_dereplicate.pl</span></a> will pool the files and dereplicate sequences
of the taxIDs that are present in both files.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/pool_and_dereplicate.pl<span class="w"> </span>-tsv1<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-tsv2<span class="w"> </span>tutorial/custom/3_dereplicate/custom_dereplicated_sequences.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/custom<span class="w"> </span>-out<span class="w"> </span>COInr_custom.tsv
</pre></div>
</div>
<p>The output is the same format as the input tsv file.</p>
<p>See details in description section: <a class="reference internal" href="reference.html#pool-and-dereplicate-reference"><span class="std std-ref">pool_and_dereplicate.pl</span></a> script.</p>
</section>
<section id="custom-database">
<h4>Custom database<a class="headerlink" href="#custom-database" title="Link to this heading"></a></h4>
<dl class="simple">
<dt>Your custom database is composed of two files:</dt><dd><ul class="simple">
<li><p>the dereplicated sequence file (COInr_custom.tsv)</p></li>
<li><p>the last version of the taxonomy file (taxonomy_updated.tsv)</p></li>
</ul>
</dd>
</dl>
<p>For simplicity, move the updated taxonomy file to the same folder as the sequence file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mv<span class="w"> </span>tutorial/custom/2_add_taxids/taxonomy_updated.tsv<span class="w"> </span>tutorial/custom/taxonomy_updated.tsv
</pre></div>
</div>
<p>This database can be further customized, or you can simply be formated to your taxonomic assignment program by the <a class="reference internal" href="reference.html#format-db-reference"><span class="std std-ref">format_db.pl</span></a> script.</p>
</section>
</section>
<section id="select-sequences-from-existing-database">
<span id="select-sequences-custom-tutorial"></span><h3>Select sequences from existing database<a class="headerlink" href="#select-sequences-from-existing-database" title="Link to this heading"></a></h3>
<section id="select-sequences-for-a-list-of-taxa-with-a-minimum-taxonomic-rank">
<h4>Select sequences for a list of taxa with a minimum taxonomic rank<a class="headerlink" href="#select-sequences-for-a-list-of-taxa-with-a-minimum-taxonomic-rank" title="Link to this heading"></a></h4>
<p>Sequences can be selected for a list of taxa and/or for a minimum taxonomic level (species/genus/family/order/class/phylum/kingdom/superkingdom/root)</p>
<p>The input file (<a class="reference internal" href="io.html#taxon-list-io"><span class="std std-ref">-taxon_list</span></a>) contains a list of taxa and eventually their taxIDs (see example data/example/taxon_list.tsv).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/select_taxa.pl<span class="w"> </span>-taxon_list<span class="w"> </span>data/example/taxon_list.tsv<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w">  </span>-min_taxlevel<span class="w"> </span>species<span class="w">  </span>-outdir<span class="w"> </span>tutorial/select_taxa_0<span class="w"> </span>-out<span class="w"> </span>COInr_selected.tsv
</pre></div>
</div>
<p>The main output is a <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv file</span></a> (COInr_selected.tsv).
A <span class="xref std std-ref">lineage file</span> (taxa_with_lineages.tsv) is also written for all taxa in the taxon_list to check if they are coherent with the target taxon names.</p>
<p>See details in description section: <a class="reference internal" href="reference.html#select-taxa-reference"><span class="std std-ref">select_taxa.pl</span></a> script.</p>
</section>
<section id="excluding-sequences-of-a-taxon-list">
<h4>Excluding  sequences of a taxon list<a class="headerlink" href="#excluding-sequences-of-a-taxon-list" title="Link to this heading"></a></h4>
<p>With the same script it is also possible to eliminate sequences of taxa instead of selecting them. Set the <em>negative_list</em> option to 1 to do that.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/select_taxa.pl<span class="w"> </span>-taxon_list<span class="w"> </span>data/example/taxon_list.tsv<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w">  </span>-min_taxlevel<span class="w"> </span>species<span class="w"> </span>-outdir<span class="w"> </span>tutorial/select_taxa_1<span class="w"> </span>-out<span class="w"> </span>COInr_reduced.tsv<span class="w"> </span>-negative_list<span class="w"> </span><span class="m">1</span>
</pre></div>
</div>
<p>See details in description section: <a class="reference internal" href="reference.html#select-taxa-reference"><span class="std std-ref">select_taxa.pl</span></a> script.</p>
</section>
</section>
<section id="select-region">
<span id="select-region-custom-tutorial"></span><h3>Select region<a class="headerlink" href="#select-region" title="Link to this heading"></a></h3>
<p>Sequences can be trimmed to a specific region of the COI gene by the <a class="reference internal" href="reference.html#select-region-reference"><span class="std std-ref">select_region.pl</span></a> script.
To define the region, you can either give a fasta file with sequences trimmed to the region of interest,
or you can detect it automatically by e-pcr.</p>
<section id="select-region-using-the-e-pcr-option">
<span id="select-region-e-pcr-custom-tutorial"></span><h4>Select region using the e_pcr option<a class="headerlink" href="#select-region-using-the-e-pcr-option" title="Link to this heading"></a></h4>
<p>The primers used in this example are amplifying a Leray fragment (ca. 313 bp of the second half of the barcode region).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/select_region.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/select_region/ePCR<span class="w"> </span>-e_pcr<span class="w"> </span><span class="m">1</span><span class="w"> </span>-fw<span class="w"> </span>GGNTGAACNGTNTAYCCNCC<span class="w"> </span>-rv<span class="w"> </span>TAWACTTCDGGRTGNCCRAARAAYCA<span class="w"> </span>-trim_error<span class="w"> </span><span class="m">0</span>.3<span class="w"> </span>-min_amplicon_length<span class="w"> </span><span class="m">280</span><span class="w"> </span>-max_amplicon_length<span class="w"> </span><span class="m">345</span><span class="w"> </span>-min_overlap<span class="w"> </span><span class="m">20</span><span class="w"> </span>-tcov<span class="w"> </span><span class="m">0</span>.8<span class="w"> </span>-identity<span class="w"> </span><span class="m">0</span>.7
</pre></div>
</div>
</section>
<section id="select-region-using-the-bait-fas-option">
<span id="select-region-bait-fas-custom-tutorial"></span><h4>Select region using the bait_fas option<a class="headerlink" href="#select-region-using-the-bait-fas-option" title="Link to this heading"></a></h4>
<p>Using the <em>e_pcr</em> option is an easy way to produce some sequences trimmed to the target region,
and they can be used as a database to align all other sequences to them.
However, if the parameters of the e_pcr are relaxed, it can produce some false positives.
An alternative solution is to use a small, taxonomically divers fasta file, with sequences already trimmed to the target region
(-<em>bait_fas</em> option).
An example of such a file is given in the data directory (data/one_seq_per_order_658.fas).
It contains one sequence for each taxonomic order among the taxa that have a compete mitochondrial genome available in GenBank.
Sequences are trimmed to the approximately 658 bp (depending on the taxon) barcode fragment of the COI gene.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/select_region.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/select_region/bait_fas<span class="w"> </span>-e_pcr<span class="w"> </span><span class="m">0</span><span class="w"> </span>-bait_fas<span class="w"> </span>data/one_seq_per_order_658.fas<span class="w"> </span>-tcov<span class="w"> </span><span class="m">0</span>.8<span class="w"> </span>-identity<span class="w"> </span><span class="m">0</span>.7
</pre></div>
</div>
<p>See details in description section: <a class="reference internal" href="reference.html#select-region-reference"><span class="std std-ref">select_region.pl</span></a> script.</p>
</section>
</section>
<section id="format-database">
<span id="format-db-custom-tutorial"></span><h3>Format database<a class="headerlink" href="#format-database" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>Format the database to one of the following formats</dt><dd><ul class="simple">
<li><p>qiime</p></li>
<li><p>rdp</p></li>
<li><p>full</p></li>
<li><p>blast</p></li>
<li><p>vtam</p></li>
<li><p>sintax</p></li>
</ul>
</dd>
</dl>
<p><strong>qiime</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_db.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outfmt<span class="w"> </span>qiime<span class="w"> </span>-outdir<span class="w"> </span>COInr/qiime<span class="w"> </span>-out<span class="w"> </span>COInr_qiime
</pre></div>
</div>
<p><strong>rdp</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_db.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outfmt<span class="w"> </span>rdp<span class="w"> </span>-outdir<span class="w"> </span>COInr/rdp<span class="w"> </span>-out<span class="w"> </span>COInr_rdp
</pre></div>
</div>
<p>You should use the rdp_calssifier or qiime’s feature-classifier to train the database using the output files of this script if you have used the rdp or qiime options.</p>
<p><strong>full</strong></p>
<p>The full option, gives a <a class="reference internal" href="io.html#full-tsv-io"><span class="std std-ref">tsv file</span></a> with seqIDs, ranked lineages, taxIDs for each sequence, and this is a very easy-to-parse, complete file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_db.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outfmt<span class="w"> </span>full<span class="w"> </span>-outdir<span class="w"> </span>COInr/full<span class="w"> </span>-out<span class="w"> </span>COInr_full
</pre></div>
</div>
<p><strong>sintax</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_db.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outfmt<span class="w"> </span>sintax<span class="w"> </span>-outdir<span class="w"> </span>COInr/sintax<span class="w"> </span>-out<span class="w"> </span>COInr_sintax
</pre></div>
</div>
<p><strong>blast</strong></p>
<p>For making a BLAST database, the taxonomy file is not necessary and the indexed files in the output folder are ready to use.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_db.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-outfmt<span class="w"> </span>blast<span class="w"> </span>-outdir<span class="w"> </span>COInr/blast<span class="w"> </span>-out<span class="w"> </span>COInr_blast
</pre></div>
</div>
<p><strong>vtam</strong></p>
<p>The vtam option produces a BLAST database and a taxonomy file adapted to <a class="reference external" href="https://github.com/aitgon/vtam">VTAM</a> .</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_db.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outfmt<span class="w"> </span>vtam<span class="w"> </span>-outdir<span class="w"> </span>COInr/vtam<span class="w"> </span>-out<span class="w"> </span>COInr_vtam
</pre></div>
</div>
<p>See details in description section: <a class="reference internal" href="reference.html#format-db-reference"><span class="std std-ref">format_db.pl</span></a> script.</p>
</section>
<section id="chaining-steps-to-make-a-custom-database">
<span id="chained-custom-tutorial"></span><h3>Chaining steps to make a custom database<a class="headerlink" href="#chaining-steps-to-make-a-custom-database" title="Link to this heading"></a></h3>
<p>In the above examples, we have started from the COInr database. However, you can chain the different commands.</p>
<dl class="simple">
<dt>Bellow, I will show you how to create a database with the following characteristics:</dt><dd><ul class="simple">
<li><p>Eukaryota sequences</p></li>
<li><p>Excluding insects</p></li>
<li><p>Enriched with custom sequences</p></li>
<li><p>Sequences assigned at least to genus level</p></li>
<li><p>Trimmed to the Leray fragment (ca. 313 nt of the second half of the barcode region) of the COI gene (keep sequences if cover at least 90% of the target region)</p></li>
<li><p>rdp_classifier format</p></li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt><dd><ul class="simple">
<li><p>It is a good idea to start with steps that are relatively quick and reduce the size of the database.</p></li>
<li><p>Since, over 70% of the sequences are from Insecta in COInr, we will start by eliminating them.</p></li>
<li><p>The custom sequences are all Non-Insect Eukaryotes, so we can add custom sequences to the reduced dataset. Otherwise, we should have started by adding custom sequences. This solution is also fine, but gives large intermediate files.</p></li>
<li><p>The selection of the target region is the most computationally intensive, and the more diverse the dataset, the less precise it is. So it is preferable to do this at the end of the pipeline.</p></li>
</ul>
</dd>
</dl>
<section id="exclude-insecta-and-sequences-with-resolution-lower-than-genus">
<span id="exclude-insecta-tutorial"></span><h4>Exclude Insecta and sequences with resolution lower than genus<a class="headerlink" href="#exclude-insecta-and-sequences-with-resolution-lower-than-genus" title="Link to this heading"></a></h4>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/select_taxa.pl<span class="w"> </span>-taxon_list<span class="w"> </span>data/example/taxon_list_insecta.tsv<span class="w"> </span>-tsv<span class="w"> </span>COInr/COInr.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w">  </span>-min_taxlevel<span class="w"> </span>genus<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/1_noInsecta<span class="w"> </span>-out<span class="w"> </span>COInr_noIns.tsv<span class="w"> </span>-negative_list<span class="w"> </span><span class="m">1</span>
</pre></div>
</div>
</section>
<section id="keep-eukaryota">
<span id="keep-eukaryota-tutorial"></span><h4>Keep Eukaryota<a class="headerlink" href="#keep-eukaryota" title="Link to this heading"></a></h4>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/select_taxa.pl<span class="w"> </span>-taxon_list<span class="w"> </span>data/example/taxon_list_eukaryota.tsv<span class="w"> </span>-tsv<span class="w"> </span>tutorial/chained/1_noInsecta/COInr_noIns.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/2_Eukaryota<span class="w"> </span>-out<span class="w"> </span>COInr_noIns_Euk.tsv
</pre></div>
</div>
</section>
<section id="add-custom-sequences">
<span id="add-custom-chained-tutorial"></span><h4>Add custom sequences<a class="headerlink" href="#add-custom-sequences" title="Link to this heading"></a></h4>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_custom.pl<span class="w"> </span>-custom<span class="w"> </span>data/example/my_sequences.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/3_add_custom/1_format
</pre></div>
</div>
<p>Check and format the custom_lineages.tsv and make custom_lineages_verified.tsv as in <a class="reference internal" href="#add-custom-sequences-tutorial"><span class="std std-ref">Add custom sequences to a database</span></a> section.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/add_taxids.pl<span class="w"> </span>-lineages<span class="w"> </span>data/example/custom_lineages_verified.tsv<span class="w"> </span>-sequences<span class="w"> </span>tutorial/chained/3_add_custom/1_format/custom_sequences.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr/taxonomy.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/3_add_custom/2_add_taxids

perl<span class="w"> </span>scripts/dereplicate.pl<span class="w"> </span>-tsv<span class="w"> </span>tutorial/chained/3_add_custom/2_add_taxids/sequences_with_taxIDs.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/3_add_custom/3_dereplicate<span class="w"> </span>-out<span class="w"> </span>custom_dereplicated_sequences.tsv
</pre></div>
</div>
<p>Add the formatted, dereplicated custom sequences to the sequences in COInr_noIns_Euk.tsv</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/pool_and_dereplicate.pl<span class="w"> </span>-tsv1<span class="w"> </span>tutorial/chained/2_Eukaryota/COInr_noIns_Euk.tsv<span class="w"> </span>-tsv2<span class="w"> </span>tutorial/chained/3_add_custom/3_dereplicate/custom_dereplicated_sequences.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/3_add_custom<span class="w"> </span>-out<span class="w"> </span>COInr_noIns_Euk_custom.tsv

mv<span class="w"> </span>tutorial/chained/3_add_custom/2_add_taxids/taxonomy_updated.tsv<span class="w"> </span>tutorial/chained/3_add_custom/taxonomy_updated.tsv
</pre></div>
</div>
</section>
<section id="keep-only-sequences-with-genus-or-higher-resolution">
<span id="keep-genus-tutorial"></span><h4>Keep only sequences with genus or higher resolution<a class="headerlink" href="#keep-only-sequences-with-genus-or-higher-resolution" title="Link to this heading"></a></h4>
<p>We have eliminated sequences with lower than genus resolution from COInr in the first step (-min_taxlevel genus).
However, among the custom sequences we had a sequence with an unknown genus.
So let’s redo the selection for a minimum taxonomic level.</p>
<p>Yes, you are right! We could have just avoided adding that sequence to the database in the previous step.
But if you have many custom sequences, you might just be lazy to check the custom sequences manually,
and in that case you can use mkCOInr to this for you.</p>
<p><strong>Attention</strong>: From now on, we have to use the updated taxonomy file (taxonomy_updated.tsv), since some of the taxa of the custom sequences might not be in the original taxonomy.tsv file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/select_taxa.pl<span class="w"> </span>-tsv<span class="w"> </span>tutorial/chained/3_add_custom/COInr_noIns_Euk_custom.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>tutorial/chained/3_add_custom/taxonomy_updated.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/4_genus<span class="w"> </span>-out<span class="w"> </span>COInr_noIns_Euk_custom_genus.tsv
</pre></div>
</div>
</section>
<section id="trim-to-leray-region">
<span id="trim-to-leray-tutorial"></span><h4>Trim to Leray region<a class="headerlink" href="#trim-to-leray-region" title="Link to this heading"></a></h4>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/select_region.pl<span class="w"> </span>-tsv<span class="w"> </span>tutorial/chained/4_genus/COInr_noIns_Euk_custom_genus.tsv<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/5_select_region<span class="w"> </span>-e_pcr<span class="w"> </span><span class="m">1</span><span class="w"> </span>-fw<span class="w"> </span>GGNTGAACNGTNTAYCCNCC<span class="w"> </span>-rv<span class="w"> </span>TAWACTTCDGGRTGNCCRAARAAYCA<span class="w"> </span>-trim_error<span class="w"> </span><span class="m">0</span>.3<span class="w"> </span>-min_amplicon_length<span class="w"> </span><span class="m">280</span><span class="w"> </span>-max_amplicon_length<span class="w"> </span><span class="m">345</span><span class="w"> </span>-min_overlap<span class="w"> </span><span class="m">20</span><span class="w"> </span>-tcov<span class="w"> </span><span class="m">0</span>.9<span class="w"> </span>-identity<span class="w"> </span><span class="m">0</span>.7
</pre></div>
</div>
</section>
<section id="format-for-rdp-classifier">
<span id="format-rdp-chained-tutorial"></span><h4>Format for RDP_classifier<a class="headerlink" href="#format-for-rdp-classifier" title="Link to this heading"></a></h4>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_db.pl<span class="w"> </span>-tsv<span class="w"> </span>tutorial/chained/5_select_region/trimmed.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>tutorial/chained/3_add_custom/taxonomy_updated.tsv<span class="w"> </span>-outfmt<span class="w"> </span>rdp<span class="w"> </span>-outdir<span class="w"> </span>tutorial/chained/6_rdp<span class="w"> </span>-out<span class="w"> </span>COInr_customized
</pre></div>
</div>
</section>
</section>
</section>
<section id="create-coinr-from-bold-and-ncbi">
<span id="create-coinr-tutorial"></span><h2>Create COInr from BOLD and NCBI<a class="headerlink" href="#create-coinr-from-bold-and-ncbi" title="Link to this heading"></a></h2>
<p>The following steps describe how COInr database (available at <a class="reference external" href="https://doi.org/10.5281/zenodo.6555985">Zenodo</a> ) was produced.</p>
<section id="download-ncbi-taxonomy">
<span id="download-ncbi-taxonomy-tutorial"></span><h3>Download NCBI taxonomy<a class="headerlink" href="#download-ncbi-taxonomy" title="Link to this heading"></a></h3>
<p>Download NCBI taxonomy dmp file and create <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>mkCOInr
perl<span class="w"> </span>scripts/download_taxonomy.pl<span class="w"> </span>-outdir<span class="w"> </span>COInr_new/taxonomy
</pre></div>
</div>
<p>See details in description section: <a class="reference internal" href="reference.html#download-taxonomy-reference"><span class="std std-ref">download_taxonomy.pl</span></a> script.</p>
</section>
<section id="ncbi-sequences">
<span id="ncbi-sequences-tutorial"></span><h3>NCBI sequences<a class="headerlink" href="#ncbi-sequences" title="Link to this heading"></a></h3>
<section id="download-ncbi-sequences">
<h4>Download NCBI sequences<a class="headerlink" href="#download-ncbi-sequences" title="Link to this heading"></a></h4>
<p>The following command will download Coding DNA Sequence (CDS) fasta files of all sequences with COI, CO1, COXI or COX1 in the title lines and complete mitochondrial genomes.
It takes several hours (days) to run this command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nsdpy<span class="w"> </span>-r<span class="w"> </span><span class="s2">&quot;COI OR COX1 OR CO1 OR COXI OR (complete[Title] AND genome[Title] AND Mitochondrion[Filter])&quot;</span><span class="w"> </span>-T<span class="w"> </span>-v<span class="w"> </span>--cds
</pre></div>
</div>
<p>The results are found in the NSDPY_results/yyyy-mm-dd_hh-mm-ss folder.</p>
<p>The sequences.fasta file contains all CDS sequences. Sequences are correctly oriented but should still be filtered to keep only COI sequences.
TaxIDs.txt contains the sequenceIDs and the TaxIDs.</p>
<p>Move the results of nsdpy to the COInr_new/ncbi/download directory and clean up the directory.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir<span class="w"> </span>-p<span class="w"> </span>COInr_new/ncbi
mv<span class="w"> </span>NSDPY_results/yyyy-mm-dd_hh-mm-ss<span class="w"> </span>COInr_new/ncbi/download
mv<span class="w"> </span>report.tsv<span class="w"> </span>COInr_new/ncbi/download
rmdir<span class="w"> </span>NSDPY_results
</pre></div>
</div>
</section>
<section id="format-ncbi-sequences">
<h4>Format NCBI sequences<a class="headerlink" href="#format-ncbi-sequences" title="Link to this heading"></a></h4>
<dl class="simple">
<dt>The <a class="reference internal" href="reference.html#format-ncbi-reference"><span class="std std-ref">format_ncbi.pl</span></a> script will</dt><dd><ul class="simple">
<li><p>Select COI sequences and clean them.</p></li>
<li><p>Eliminate identical sequences of the same taxID.</p></li>
<li><p>Clean tax names and taxids.</p></li>
</ul>
</dd>
</dl>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_ncbi.pl<span class="w"> </span>-cds<span class="w"> </span>COInr_new/ncbi/download/sequences.fasta<span class="w"> </span>-taxids<span class="w"> </span>COInr_new/ncbi/download/TaxIDs.txt<span class="w"> </span>-taxonomy<span class="w"> </span>COInr_new/taxonomy/taxonomy.tsv<span class="w"> </span>-outdir<span class="w"> </span>COInr_new/ncbi/format
</pre></div>
</div>
<p>The major output is a <a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequence tsv file with taxIDs</span></a>.</p>
<p>See details in description section: <a class="reference internal" href="reference.html#format-ncbi-reference"><span class="std std-ref">format_ncbi.pl</span></a> script.</p>
</section>
<section id="dereplicate-ncbi-sequences">
<h4>Dereplicate NCBI sequences<a class="headerlink" href="#dereplicate-ncbi-sequences" title="Link to this heading"></a></h4>
<p>Eliminate sequences that are substring of another sequence of the same <a class="reference internal" href="glossary.html#taxid-glossary"><span class="std std-ref">taxID</span></a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/dereplicate.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr_new/ncbi/format/ncbi_sequences.tsv<span class="w"> </span>-outdir<span class="w"> </span>COInr_new/ncbi/dereplicate<span class="w"> </span>-out<span class="w"> </span>ncbi_dereplicated_sequences.tsv
</pre></div>
</div>
<p>The output is the same format as the input tsv file.</p>
<p>See details in description section: <a class="reference internal" href="reference.html#dereplicate-reference"><span class="std std-ref">dereplicate.pl</span></a> script.</p>
</section>
</section>
<section id="bold-sequences">
<span id="bold-sequences-tutorial"></span><h3>BOLD sequences<a class="headerlink" href="#bold-sequences" title="Link to this heading"></a></h3>
<section id="download-bold-sequences">
<h4>Download BOLD sequences<a class="headerlink" href="#download-bold-sequences" title="Link to this heading"></a></h4>
<p>The <a class="reference internal" href="reference.html#download-bold-reference"><span class="std std-ref">download_bold.pl</span></a> script is deprecated. The BOLD API used in download_bold.pl do not allow anymore to download large data files.</p>
<p>It is possible, however, to download all public sequences as a data package from
<a class="reference external" href="https://www.boldsystems.org/index.php/datapackages">https://www.boldsystems.org/index.php/datapackages</a>.
You need to have a BOLD account for downloading the data package in (tar.gz compressed) format,
that contains a TSV file with sequences, taxonomic lineages and other metadata. This uncompressed file will be the input of format_bold_package.pl.</p>
</section>
<section id="format-bold-sequences">
<h4>Format BOLD sequences<a class="headerlink" href="#format-bold-sequences" title="Link to this heading"></a></h4>
<dl class="simple">
<dt>The <a class="reference internal" href="reference.html#format-bold-package-reference"><span class="std std-ref">format_bold_package.pl</span></a> script will</dt><dd><ul class="simple">
<li><p>Select COI sequences and clean them</p></li>
<li><p>Eliminate identical sequences of the same lineage</p></li>
<li><p>Clean lineages and make a list with corresponding sequenceIDs</p></li>
</ul>
</dd>
</dl>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/format_bold_package.pl<span class="w"> </span>-bold_data<span class="w"> </span>COInr_new/bold/download/BOLD_Public.26-Apr-2024.tsv<span class="w"> </span>-outdir<span class="w"> </span>COInr_new/bold/format
</pre></div>
</div>
<dl class="simple">
<dt>The major output is the following:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-without-taxid-io"><span class="std std-ref">bold_sequences.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#lineage-tsv-without-taxid-io"><span class="std std-ref">bold_lineages.tsv</span></a> (all identical lineages are pooled into a same line)</p></li>
</ul>
</dd>
</dl>
<p>See details in description section: <a class="reference internal" href="reference.html#format-bold-package-reference"><span class="std std-ref">format_bold_package.pl</span></a> script.</p>
</section>
<section id="add-taxids-to-bold-sequences">
<h4>Add taxIDs to BOLD sequences<a class="headerlink" href="#add-taxids-to-bold-sequences" title="Link to this heading"></a></h4>
<dl class="simple">
<dt>For each lineage the <a class="reference internal" href="reference.html#add-taxids-reference"><span class="std std-ref">add_taxids.pl</span></a> script will</dt><dd><ul class="simple">
<li><p>Find an existing <a class="reference internal" href="glossary.html#taxid-glossary"><span class="std std-ref">taxID</span></a> at the lowest level possible. TaxIDs can be either from NCBI, or negative taxID already present in <a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy.tsv</span></a>.</p></li>
<li><p>Add new arbitrary (negative) taxIDs to taxa, that are not yet in taxonomy.tsv</p></li>
<li><p>Link each new taxID to existing one as a child and include info to the updated taxonomy file</p></li>
<li><p>Update the input taxonomy file</p></li>
</ul>
</dd>
</dl>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/add_taxids.pl<span class="w"> </span>-lineages<span class="w"> </span>COInr_new/bold/format/bold_lineages.tsv<span class="w"> </span>-sequences<span class="w"> </span>COInr_new/bold/format/bold_sequences.tsv<span class="w"> </span>-taxonomy<span class="w"> </span>COInr_new/taxonomy/taxonomy.tsv<span class="w"> </span>-outdir<span class="w"> </span>COInr_new/bold/add_taxids
</pre></div>
</div>
<dl class="simple">
<dt>The main output files are the following:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="io.html#sequence-tsv-with-taxid-io"><span class="std std-ref">sequences_with_taxIDs.tsv</span></a></p></li>
<li><p><a class="reference internal" href="io.html#taxonomy-io"><span class="std std-ref">taxonomy_updated.tsv</span></a></p></li>
</ul>
</dd>
</dl>
<p>See details in description section: <a class="reference internal" href="reference.html#add-taxids-reference"><span class="std std-ref">add_taxids.pl</span></a> script.</p>
</section>
<section id="dereplicate-bold-sequences">
<h4>Dereplicate BOLD sequences<a class="headerlink" href="#dereplicate-bold-sequences" title="Link to this heading"></a></h4>
<p>Eliminate sequences that are substring of another sequence of the same taxID.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/dereplicate.pl<span class="w"> </span>-tsv<span class="w"> </span>COInr_new/bold/add_taxids/sequences_with_taxIDs.tsv<span class="w"> </span>-outdir<span class="w"> </span>COInr_new/bold/dereplicate<span class="w"> </span>-out<span class="w"> </span>bold_dereplicated_sequences.tsv
</pre></div>
</div>
<p>The output is the same format as the input tsv file.</p>
<p>See details in description section: <a class="reference internal" href="reference.html#dereplicate-reference"><span class="std std-ref">dereplicate.pl</span></a> script.</p>
</section>
</section>
<section id="pool-and-dereplicate-tutorial">
<span id="id2"></span><h3>Pool and dereplicate datasets<a class="headerlink" href="#pool-and-dereplicate-tutorial" title="Link to this heading"></a></h3>
<p>Use the dereplicated sequence files from BOLD and NCBI.
The <a class="reference internal" href="reference.html#pool-and-dereplicate-reference"><span class="std std-ref">pool_and_dereplicate.pl</span></a> script will pool the files and dereplicate sequences of a taxID that are present in both files.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl<span class="w"> </span>scripts/pool_and_dereplicate.pl<span class="w"> </span>-tsv1<span class="w"> </span>COInr_new/bold/dereplicate/bold_dereplicated_sequences.tsv<span class="w"> </span>-tsv2<span class="w"> </span>COInr_new/ncbi/dereplicate/ncbi_dereplicated_sequences.tsv<span class="w"> </span>-outdir<span class="w"> </span>COInr_new<span class="w"> </span>-out<span class="w"> </span>COInr.tsv
</pre></div>
</div>
<p>The output is the same format as the input tsv file.</p>
<p>See details in description section: <a class="reference internal" href="reference.html#pool-and-dereplicate-reference"><span class="std std-ref">pool_and_dereplicate.pl</span></a> script.</p>
<p><strong>Move the taxonomy file to the same directory</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mv<span class="w"> </span>COInr_new/bold/add_taxids/taxonomy_updated.tsv<span class="w"> </span>COInr_new/taxonomy.tsv
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="run.html" class="btn btn-neutral float-left" title="Run mkCOInr" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="reference.html" class="btn btn-neutral float-right" title="Description" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Emese Meglécz.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>